extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,430 @@
+                                               TIME_TO_TICKS(SHT3x_MEAS_DURATION_REP_LOW) };
+
+#if defined(SHT3x_DEBUG_LEVEL_2)
+#define debug(s, f, ...) printf("%s %s: " s "\n", "SHT3x", f, ## __VA_ARGS__)
+#define debug_dev(s, f, d, ...) printf("%s %s: bus %d, addr %02x - " s "\n", "SHT3x", f, d->bus, d->addr, ## __VA_ARGS__)
+#else
+#define debug(s, f, ...)
+#define debug_dev(s, f, d, ...)
+#endif
+
+#if defined(SHT3x_DEBUG_LEVEL_1) || defined(SHT3x_DEBUG_LEVEL_2)
+#define error(s, f, ...) printf("%s %s: " s "\n", "SHT3x", f, ## __VA_ARGS__)
+#define error_dev(s, f, d, ...) printf("%s %s: bus %d, addr %02x - " s "\n", "SHT3x", f, d->bus, d->addr, ## __VA_ARGS__)
+#else
+#define error(s, f, ...)
+#define error_dev(s, f, d, ...)
+#endif
+
+/** Forward declaration of function for internal use */
+
+static bool sht3x_is_measuring  (sht3x_sensor_t*);
+static bool sht3x_send_command  (sht3x_sensor_t*, uint16_t);
+static bool sht3x_read_data     (sht3x_sensor_t*, uint8_t*,  uint32_t);
+static bool sht3x_get_status    (sht3x_sensor_t*, uint16_t*);
+static bool sht3x_reset         (sht3x_sensor_t*);
+
+static uint8_t crc8 (uint8_t data[], int len);
+
+/** ------------------------------------------------ */
+
+bool sht3x_init_driver()
+{
+    return true;
+}
+
+
+sht3x_sensor_t* sht3x_init_sensor(uint8_t bus, uint8_t addr)
+{
+    sht3x_sensor_t* dev;
+
+    if ((dev = malloc (sizeof(sht3x_sensor_t))) == NULL)
+        return NULL;
+    
+    // inititalize sensor data structure
+    dev->bus  = bus;
+    dev->addr = addr;
+    dev->mode = single_shot;
+    dev->repeatability = high;
+    dev->meas_start_time = 0;
+    dev->meas_started = false;
+    dev->meas_first = false;
+
+    uint16_t status;
+
+    // reset the sensor
+    if (!sht3x_reset(dev))
+    {
+        error_dev ("could not reset the sensor", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+    
+    // check again the status after clear status command
+    if (!sht3x_get_status(dev, &status))
+    {
+        error_dev ("could not get sensor status", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+    
+    debug_dev ("sensor initialized", __FUNCTION__, dev);
+    return dev;
+}
+
+
+bool sht3x_start_measurement (sht3x_sensor_t* dev, sht3x_mode_t mode)
+{
+    if (!dev) return false;
+    
+    dev->error_code = SHT3x_OK;
+    dev->mode = mode;
+    
+    // start measurement according to selected mode and return an duration estimate
+    if (!sht3x_send_command(dev, SHT3x_MEASURE_CMD[dev->mode][dev->repeatability]))
+    {
+        error_dev ("could not send start measurment command", __FUNCTION__, dev);
+        dev->error_code |= SHT3x_SEND_MEAS_CMD_FAILED;
+        return false;
+    }
+
+    dev->meas_start_time = sdk_system_get_time ();
+    dev->meas_started = true;
+    dev->meas_first = true;
+
+    return true;
+}
+
+
+uint8_t sht3x_get_measurement_duration (sht3x_sensor_t* dev)
```

Comment:
I would just pass in the 'repeatability' argument. That way it is clear to the caller that this depends only on the 'repeatability' argument and is a constant that they can hoist and that they need not lock the i2c bus for this call.

Reply:
@ourairquality 
I understand your intention. Unfortunately, my understanding of this function was more like "Give me the duration for the current measurement configuration of the sensor" and not "What would the duration be if the sensor would measure with a certain configuration". Exactly that was the reason why function ```sht3x_start_measurement``` initially returned the duration.

I know that you just only try to optimize everything that is constant. But, the function call costs almost nothing. It is just the call and a table lookup. Indeed it could be realized as macro.

Furthermore, this question is related to the question whether to have a repeatability parameter in ```sht3x_start_measurement``` or not. If ```sht3x_start_measurement``` would have this parameter, the user already knows something about repeatability and it would be ok. Without this parameter, the function would be confusing.

At the moment it is a bit frustrating to change again and again the interface and to change the documentation every time. Latter part is the most time consuming part. Therefore I would like to come to an interface code freeze. The code behind the interface can changed anytime later, but the interface must never change.

Reply:
@ourairquality 
What if the user gets the duration for low repeatability and then starts a measurement at high repeatability? I know, this would be a coding error. But the driver is not fault tolerant. Therefore, my idea was that the user gets always the duration the sensor need for the current configuration.

Reply:
@ourairquality 
In my first approach I didn't plan to give the user the chance to select the repeatability. I planned to perform measurement always with high repeatability.

Reply:
I don't mind if the driver lands as-is. I don't want to impose my views on this design and that is not necessary when it is an extra component that people can optionally use. We probably just have different views on coding, my preference is minimal abstraction and for run time checks to be a compile time option.